### PR TITLE
Introduce protobuf serialization/deserialization support for transport requests/responses

### PR DESCRIPTION
**Pull request by VachaShah**
### Description
This is the first PR for adding support for protobuf serialization/deserialization coming from the POC and draft PR #9097. The idea is to add incremental changes (bottom-up) for the support to use protobuf for cat nodes API. In this PR, the following changes are included:

* `ProtobufWriteable` - A version of Writeable where the reader reads from a byte array and writer writes to a java OutputStream. This is needed to serialize and deserialize to and from proto messages. There will be other changes in the OutboundHandler and InboundHandler to transfer bytes related to proto messages over the wire.
* `TransportRequest` and `TransportResponse` support ProtobufWriteable.
* `TaskId` is converted into a proto message. 
* `TaskIdTests` are moved to `libs/core` module since `TaskId` has been moved there.

### Related Issues
#6844 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
I like it. A couple of ideas/comments.

1. I think it's worth starting to document the new protobuf-based protocol from the start. This PR could contain some README content on how to enable protobuf, what parts of the protocol it replaces (e.g. some extension support is protobuf only).
2. I quickly read the code, how do I exercise it in a build of OpenSearch? I think it's important to find a path in which we can have something "useful" early.
3. Where can one read about the north star of completely replacing the custom-rolled OpenSearch protocol by Protobuf?

**Comment by dblock** [comment]
@dbwiddis will have lots of opinions here ;)

File: libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java [hunk 1]
```diff
@@ -84,9 +91,11 @@ public TaskId(String taskId) {
             } catch (NumberFormatException ex) {
                 throw new IllegalArgumentException("malformed task id " + taskId, ex);
             }
+            taskIdProto = TaskIdProto.TaskId.newBuilder().setNodeId(nodeId).setId(id).build();
         } else {
             nodeId = "";
             id = -1L;
+            taskIdProto = TaskIdProto.TaskId.newBuilder().setNodeId(nodeId).setId(id).build();
```
Comment (dbwiddis) [inline libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java:98]: See above on serializing "empty" information vs. never even looking at this object.

File: libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java [hunk 2]
```diff
@@ -106,6 +115,23 @@ public static TaskId readFromStream(StreamInput in) throws IOException {
         return new TaskId(nodeId, in.readLong());
     }
 
+    /**
+     * Read a {@linkplain TaskId} from a byte array. {@linkplain TaskId} has this rather than the usual constructor that takes a
+    * {@linkplain byte[]} so we can return the {@link #EMPTY_TASK_ID} without allocating.
+    */
+    public static TaskId readFromBytes(byte[] in) throws IOException {
+        TaskIdProto.TaskId taskIdProto = TaskIdProto.TaskId.parseFrom(in);
```
Comment (dbwiddis) [inline libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java:123]: This approach forces us to always parse a bytestream for the "most common" empty task ID.  

Can we have a single byte boolean (basically the "OptionalWriteable" implementation) here which directly shortcuts to `EMPTY_TASK_ID` instead of always deserializing some number of bytes > 1?

File: libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java [hunk 3]
```diff
@@ -116,6 +142,11 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeLong(id);
     }
 
+    @Override
+    public void writeTo(OutputStream out) throws IOException {
+        out.write(this.taskIdProto.toByteArray());
```
Comment (dbwiddis) [inline libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java:147]: May want to consider special-casing the empty-node-id = 0 case here if you implement it per my earlier comments.

File: libs/core/src/main/proto/tasks/TaskIdProto.proto [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+ */
+
+syntax = "proto3";
+package org.opensearch.core.tasks.proto;
+
+option java_outer_classname = "TaskIdProto";
+
+message TaskId {
+  string nodeId = 1;
+  int64 id = 2;
```
Comment (dbwiddis) [inline libs/core/src/main/proto/tasks/TaskIdProto.proto:19]: If you choose to serialize the task ID instead of using the "0" byte, consider making these `optional` parts of the message with the defaults from the empty string/-1, to accomplish the same behavior but permit serializing "nothing" rather than a string and int64 (I'm curious what a "nothing" byte stream looks like.)

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Great work so far.  Some comments/questions in-line.

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+package org.opensearch.core.common.io.stream;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+/**
+ * Implementers can be written to a {@linkplain StreamOutput} and read from a {@linkplain StreamInput}. This allows them to be "thrown
+ * across the wire" using OpenSearch's internal protocol. If the implementer also implements equals and hashCode then a copy made by
+ * serializing and deserializing must be equal and have the same hashCode. It isn't required that such a copy be entirely unchanged.
```
[the remark this answers is not in the document]
Comment (VachaShah) [reply libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java:17]: That is my bad! I will fix the comments.

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * @opensearch.internal
+ */
+public interface ProtobufWriteable {
```
[the remark this answers is not in the document]
Comment (VachaShah) [reply libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java:21]: 1. The idea is to have a similar interface to `Writeable` (`ProtobufWriteable` here for now) which will be implemented by the APIs instead of `Writeable`. 
* For the cat nodes API for which the POC is done in PR #9097, the `ClusterStateRequest`, `ClusterStateResponse` and so on will implement `ProtobufWriteable` and will be proto messages. 
* The `TransportRequest` and other common classes will continue to implement `Writeable` along with `ProtobufWriteable` since they are common classes. 
* The `OutboundHandler` which handles sending messages between nodes will also implement logic to handle proto messages which will then be converted to bytes to send over the wire and received by the `InboundHandler` thus deserializing the proto message. 
* The `OutboundMessage` will be also be a complete proto message for such cases. Reference of how the outbound message proto will look like with header as a single object: https://github.com/opensearch-project/OpenSearch/pull/9097/files#diff-9927d25e0009b6c7c9a44a8462968a10a83978bd10791b57cfa343f72f182e95.

The above is a long comment ;) but the explanation might clear a few things up. LMK what you think about this @dbwiddis!

File: libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java [hunk 7]
```diff
@@ -70,6 +76,7 @@ public TaskId(String nodeId, long id) {
     private TaskId() {
         nodeId = "";
         id = -1;
+        taskIdProto = TaskIdProto.TaskId.newBuilder().setNodeId(nodeId).setId(id).build();
```
[the remark this answers is not in the document]
Comment (VachaShah) [reply libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java:79]: See my above comment of how the outbound message will also be a proto message which will contain header, thread context request, response headers, features and action name within a single proto object.

File: libs/core/src/main/proto/tasks/TaskIdProto.proto (hunk 4 above)
Comment (VachaShah) [reply libs/core/src/main/proto/tasks/TaskIdProto.proto:19]: That makes sense! We can make the `nodeId` optional in the proto message.

**Comment by VachaShah** [comment]
> Great work so far. Some comments/questions in-line.

Thank you @dbwiddis! I have added a few responses, LMK what you think. I will keep refining the PR :)

**Comment by VachaShah** [comment]
> I like it. A couple of ideas/comments.
> 
> 1. I think it's worth starting to document the new protobuf-based protocol from the start. This PR could contain some README content on how to enable protobuf, what parts of the protocol it replaces (e.g. some extension support is protobuf only).
> 2. I quickly read the code, how do I exercise it in a build of OpenSearch? I think it's important to find a path in which we can have something "useful" early.
> 3. Where can one read about the north star of completely replacing the custom-rolled OpenSearch protocol by Protobuf?

Thank you @dblock for your comments!
1. Definitely! There will be more PRs for the same to add the support for the first API in OpenSearch (cat nodes). I will start a README for the same.
2. It will still need a few PRs to reach a point where this code can be exercised. I am adding in the changes incrementally but all the changes for the POC are in #9097. The POC ofcourse has some duplicate files since I created a parallel API to cat nodes with protobuf de/serialization for comparison.
3. I will write up an issue soon with all the work done, the next steps and what we are aiming for.

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (dbwiddis) [reply libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java:20]: > The OutboundMessage will be also be a complete proto message for such cases

Great, I peeked at that implementation, that's good all that is combined.  Let's use that as the context for my future comments.  The way I understand your intent, we will have up to 3 separate chunks of protobuf bytes:
1. The header bytes (always)
2. The bytes for TaskId (always even if empty)
3. The bytes for the TransportMessage (if any)

My comments basically boil down to two things.
1. Can (2) can be somehow folded into (1)?  If so that reduces us to 2 messages which is probably as small as we'll get.   That may not be easily done, however, and that's OK.
2. If not, let's make (2) as efficient as possible, for which I see three alternatives:
   - (My favorite) Do something similar to `OptionalWriteable` and append a boolean bit before (2).  0 means it's not there and skip it, 1 means to read the protobuf taskid.
    - Faithfully represent the existing inheritance structure and figure out how to fold the taskId into the transport message. Protobuf doesn't do inheritance but it does have composition: we could make a higher level proto for the TransportRequest that includes 2 elements: the task ID and the message.  They wouldn't be serialized until both were present; so the existing code in this PR that serializes the taskID by itself would not be needed.  (This is complex and possibly error-prone but likely the most efficient.)
   - (My least favorite) Make the bytes required for taskId a small constant (optional/missing nodeId and -1).

File: libs/core/src/main/proto/tasks/TaskIdProto.proto (hunk 4 above)
Comment (dbwiddis) [reply libs/core/src/main/proto/tasks/TaskIdProto.proto:19]: > We can make the nodeId optional in the proto message.

Note this is an alternative to the "0" shortcut which I think is a better option.   Is it allowed to have all the values optional?  What is the byte size of a protobuffer with a missing nodeId and a single -1 long, which would be the constant we'd store and use as an alternative to "0"?

**Review by dbwiddis — COMMENTED** [review]
Thanks for the clarifications.  Down to some summary comments.

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (dbwiddis) [reply libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java:20]: Did a bit more thinking about the tradeoffs here and digging into the specific benefits, and wanted to add a bit more context surrounding my main concerns here.

Overall I think this work is fantastic for enabling (arbitrarily large) TransportMessage serialization and I think that will enable great performance improvements.  My concern is getting too fine grained and serializing small pieces when we don't need to.   And I don't want to make something API when I think it's going to be a bad idea. My main objection here is the `TaskIdProto` class which you've marked API.  I don't think it should be serialized.

The current representation of the empty task ID is 1 byte, 0x00.  The protobuf representation of the "empty" task ID will take 4 bytes (either `0a 00 10 01` or `10 01 0a 00`):
 - String has an index and type byte (0x0a for index 1, length type), followed by the length byte (0)
 - Long -1 will have an index and type byte (0x10 for index 2, varint type) followed by the value (0x01 for -1)

Plus we need to pass a length byte so we know only to deserialize those 4 bytes, so we are changing the current `0x00` to `0x040a001001`.

For other task IDs, we're also adding two index bytes and a length byte.  We're saving a bit more than that in most cases due to the variable width representation but I'd argue we'd beat Protobuf in all cases by switching TaskID to use `writeVLong()` instead of `writeLong()` to squeeze out those net 2 or 3 bytes.  (We could do the same for the TcpHeader, since it contains fixed values that will never change). But really, the performance benefit is in reducing network packets, so a few bytes in the header won't impact very much for typical MTUs of 1300-1500 and are likely not worth it for a vast majority of requests.

The big benefit of protobuf is large messages, usually responses carrying a large amount of int32 or int64 data.  All the rest of this PR (particularly the responses) enables that and I'm all for it!   But I'd prefer this PR stick to the "enable messages" and not try to apply protobuf everwhere.

Protobuf is actually worse for strings (so we shouldn't serialize features and headers and the action name) and worse for collections (also not good for features and headers) and worse if you split up segments and serialize them separately (which we have to do for TCP Header / feature+action / task ID bits.).

**Comment by VachaShah** [comment]
@dbwiddis I updated the code to represent the empty taskId using an empty byte array for protobuf taskId as discussed. Let me know what you think.

File: libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java [hunk 8]
```diff
@@ -106,6 +123,22 @@ public static TaskId readFromStream(StreamInput in) throws IOException {
         return new TaskId(nodeId, in.readLong());
     }
 
+    /**
+     * Read a {@linkplain TaskId} from a byte array. {@linkplain TaskId} has this rather than the usual constructor that takes a
+    * {@linkplain byte[]} so we can return the {@link #EMPTY_TASK_ID} without allocating.
+    */
+    public static TaskId readFromBytes(byte[] in) throws IOException {
+        TaskIdProto.TaskId taskIdProto = TaskIdProto.TaskId.parseFrom(in);
+        if (!taskIdProto.hasNodeId()) {
```
Comment (dbwiddis) [inline libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java:132]: Can you help me understand this line of code as I don't see where it's defined in Java, so it may be a generated class.

My thought was that we would read/write a byte array with a VInt length before it, and the array itself gets parsed with protobuf.  If the length is 0 then we skip reading anything and return empty.

If the entire TaskId was optional (see optionalWriteable) this would essentially be the same thing.

But this looks like we are still reading some number of bytes and just determining whether the nodeId index is present; we'd still have 2 bytes for the int even if 0 bytes for the string.

Comment (VachaShah) [reply libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java:132]: The `hasNodeId` method comes from the generated class from protobuf which checks if the optional field nodeId is set or not. For line 132, I kept the same logic as `readFromStream` where if the nodeId is present, it creates a TaskId out of it else it just returns an empty taskId.

Comment (VachaShah) [reply libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java:132]: I think from what I understand, we would be able to achieve the same with keeping both nodeId and id both are optional in the proto message, so similar to the `readFromStream` where it reads the string first as the nodeId and determines if empty taskId should be returned. What do you think?

Comment (dbwiddis) [reply libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java:132]: Seems reasonable. Sorry to be so into the details here, but my main concern has been replacing a single-byte solution with anything more complex.   If making both nodeId and id optional results in a 1-byte buffer that works great.

I'm still not clear on how protobuf communicates the total bytes, though.  Looking at #6960 I see the pattern `parseFrom(in.readByteArray())` which implies a length VInt followed by that number of bytes. 

I don't see any equivalent here, or how if everything is optional it provides any signal to the reader that there's nothing to read.  Does protobuf assume reading to end-of-stream?  In that case, reading TaskId would spill over into the TransportMessage.  Or is there an end-of-stream marker?

TLDR: the savings from fewer bytes on the variable length long need to outweigh the extra bytes for field indices and length bytes and any other overhead to make serializing TaskId worth it, and I'm not sure of those numbers here.

Comment (VachaShah) [reply libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java:132]: Totally makes sense! I was thinking about this and we are only writing the bytes when the nodeId is present (https://github.com/opensearch-project/OpenSearch/pull/9737/files#diff-440ec1fd7eb50816eb2d1ec1eb201229b810efcf1939a9953afb0025ffca2f32R154) so I am thinking when we read the bytes, the byte array will have both nodeId and id or nothing at all. Would that not be similar to the current implementation? I might be totally misunderstanding it so would love to know your thoughts.

Also, currently TaskId looks like a separate proto but ultimately it will be part of the OutboundMessageProto with all the bytes and protobuf will deserialize it into various parts.

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import java.io.IOException;
+import java.io.OutputStream;
+
+/**
+ * Implementers can be written to a {@linkplain OutputStream} and read from a {@linkplain byte[]}. This allows them to be "thrown
+ * across the wire" using OpenSearch's internal protocol with protobuf serialization/de-serialization.
+ *
+ * @opensearch.internal
+ */
+public interface ProtobufWriteable {
```
Comment (Bukhtawar) [inline libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java:20]: Can we have this writeable extend `Writeable` interface?

File: libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java [hunk 10]
```diff
@@ -35,24 +35,29 @@
...
+import java.io.OutputStream;
 
 /**
  * Task id that consists of node id and id of the task on the node
  *
  * @opensearch.api
  */
 @PublicApi(since = "1.0.0")
-public final class TaskId implements Writeable {
+public final class TaskId implements Writeable, ProtobufWriteable {
```
Comment (Bukhtawar) [inline libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java:55]: Looks confusing, why do we need two interfaces?

File: libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java (hunk 8 above)
Comment (dbwiddis) [reply libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java:132]: > Totally makes sense! I was thinking about this and we are only writing the bytes when the nodeId is present (https://github.com/opensearch-project/OpenSearch/pull/9737/files#diff-440ec1fd7eb50816eb2d1ec1eb201229b810efcf1939a9953afb0025ffca2f32R154) so I am thinking when we read the bytes, the byte array will have both nodeId and id or nothing at all. Would that not be similar to the current implementation? I might be totally misunderstanding it so would love to know your thoughts.

The issue we're facing here here is that the TaskId is in the middle of the byte stream.   `StreamInput` needs to know where TaskID starts and ends so it can move on to read the bytes for the `TransportRequest`.

Take for example one of the simplest TransportRequests, the initial TCP Handshake Request:
https://github.com/opensearch-project/OpenSearch/blob/5b864c0e023647cf653300d638f60fc0136bf95b/server/src/main/java/org/opensearch/transport/TransportHandshaker.java#L192-L198

Here's what the byte stream looks like (non-protobuf):
```
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 45 53 00 00 00 31 00 00 00 00 00 00 00 08 08 08 |ES...1..........|
|00000010| 20 0b 83 00 00 00 1a 00 00 00 16 69 6e 74 65 72 | ..........inter|
|00000020| 6e 61 6c 3a 74 63 70 2f 68 61 6e 64 73 68 61 6b |nal:tcp/handshak|
|00000030| 65 00 04 a3 8e b7 41                            |e.....A         |
+--------+-------------------------------------------------+----------------+
```
The first 49 bytes (all the way up to the `65` (e) starting the last row) are all part of the `OutboundMessageRequest` header.  So we are receiving the last six bytes that we need to parse into a `TransportMessage`.

The last 5 bytes (`04 a3 8e b7 41`) are the `Writeable` payload for that `HandshakeRequest` class linked above (see `readBytesReference()`, it's a length byte followed by the 4 bytes encoding the version).

That single `00` byte in between the headers and payload is the Task ID.  It's the length of the ID string (0) which shortcuts the 8-byte id.  This is good; when writing if it's "Empty" we just write a 0 byte; when reading when we see a 0 byte we just skip it.

If you "write nothing" how will Protobuf know not to parse it?  There has to be at lease some way to indicate where the TaskID protobuf bytes start and end before we go on to read those last bytes encoding the message.

By the time you get to where `hasNodeId()` works, you've already tried to parse the message into a protobuf, but you don't know where it ends and the "version" in this handshake begins.

I think you need to do something similar to this implementation:
https://github.com/opensearch-project/OpenSearch/blob/5b864c0e023647cf653300d638f60fc0136bf95b/server/src/main/java/org/opensearch/extensions/rest/RegisterRestActionsRequest.java#L38-L47

This reads/writes the protobuf bytes wrapped inside a `ByteArray`.    You can read the byte array and if it is `StreamInput.EMPTY_BYTE_ARRAY` then you can just return the `EMPTY_TASK_ID`.  Otherwise, you take the byte array and call `parseFrom()` on it.

This preserves the 1-byte "empty" behavior we currently have and gives us a slightly shorter (due to variable length task id long) set of bytes (plus the length byte) in the case that we actually have a Task ID.

**Comment by VachaShah** [comment]
I discussed with Dan and I am closing this PR in favor of a new one with starting top down from the POC #9097 as it will make better sense as a PR and give proper context. We also discussed that we can leave TaskId alone for now and focus on the larger request/response classes to get immediate benefits.

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java (hunk 9 above)
Comment (rishabhmaurya) [reply libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java:20]: I agree with @Bukhtawar. These changes should be least intrusive and should use existing interfaces. It would enable us to experiment with other protocols as well.
